aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -40,47 +40,48 @@ pub struct GroupWrite {
     /// derive metadata_op will be validated during parallel execution to make sure
     /// it is correct, and the bytes will be replaced after the transaction is committed
     /// with correct serialized group update to obtain storage WriteOp.
-    metadata_op: WriteOp,
+    pub metadata_op: WriteOp,
     /// Updates to individual group members. WriteOps are 'legacy', i.e. no metadata.
     /// If the metadata_op is a deletion, all (correct) inner_ops should be deletions,
     /// and if metadata_op is a creation, then there may not be a creation inner op.
     /// Not vice versa, e.g. for deleted inner ops, other untouched resources may still
     /// exist in the group. Note: During parallel block execution, due to speculative
     /// reads, this invariant may be violated (and lead to speculation error if observed)
     /// but guaranteed to fail validation and lead to correct re-execution in that case.
-    inner_ops: BTreeMap<StructTag, (WriteOp, Option<Arc<MoveTypeLayout>>)>,
+    pub inner_ops: BTreeMap<StructTag, (WriteOp, Option<Arc<MoveTypeLayout>>)>,
+    /// Group size as used for gas charging, None if (metadata_)op is Deletion.
+    maybe_group_op_size: Option<u64>,
 }
 
 impl GroupWrite {
     /// Creates a group write and ensures that the format is correct: in particular,
     /// sets the bytes of a non-deletion metadata_op by serializing the provided size,
     /// and ensures inner ops do not contain any metadata.
     pub fn new(
-        mut metadata_op: WriteOp,
+        metadata_op: WriteOp,
+        inner_ops: Vec<(StructTag, (WriteOp, Option<Arc<MoveTypeLayout>>))>,
         group_size: u64,
-        inner_ops: BTreeMap<StructTag, (WriteOp, Option<Arc<MoveTypeLayout>>)>,
     ) -> Self {
-        for (v, _layout) in inner_ops.values() {
-            assert_none!(v.metadata());
+        for (_tag, (v, _layout)) in &inner_ops {
+            assert_none!(v.metadata(), "Group inner ops must have no metadata");
```

Comment:
based on @zekun000's comments on included PR: let's also assert that metadata op has no bytes or empty bytes (no longer encoding the size there).

Reply:
Done.

Reply:
When I added the assert statement, 10 unit tests in change_set.rs are failing. Is it the issue with the assert statement or the unit tests?

Reply:
pushed the fix (unit tests were putting some bytes in there) and uncommented the assert, this should be resolved.